src/Transform/ONNX/ONNXDimAnalysis.cpp

```diff
@@ -0,0 +1,588 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------- ONNXDimAnalysis.cpp - ONNX Dimension Analysis ---------------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file implements an analysis on unknown dimensions in ONNX ops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "llvm/Support/Debug.h"
+
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOpsHelper.hpp"
+#include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+#include "src/Pass/Passes.hpp"
+#include "src/Support/TypeUtilities.hpp"
+#include "src/Transform/ONNX/ONNXDimAnalysis.hpp"
+
+#define DEBUG_TYPE "dim_analysis"
+
+using namespace mlir;
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Helper functions for dimension analysis.
+//===----------------------------------------------------------------------===//
+
+/// Check if two sets are interleaved or not.
```

Comment:
Interleave, do you mean do the the two sets intersect? A set based name would be clearer, maybe `nonEmptyIntersection`